pkg/labels/v2/label.go

```diff
@@ -0,0 +1,232 @@
+	"fmt"
+	"net/netip"
+	"strings"
+	"unique"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// Label is a unique source, key and value tuple.
+type Label unique.Handle[labelRep]
+
+// NewLabel returns a new label. Labels created with the same
+// source, key and value will share a single allocation.
+func NewLabel(key, value, source string) Label {
+	return NewCIDRLabel(key, value, source, nil)
+}
+
+func NewCIDRLabel(key, value, source string, cidr *netip.Prefix) Label {
+	skv := source + ":" + key
+	if value != "" {
+		skv += "=" + value
+	}
+	return Label(unique.Make(labelRep{
+		skv:  skv,
+		kpos: uint16(len(source) + 1),
+		vpos: uint16(len(source) + 1 + len(key) + 1),
+		cidr: cidr,
+	}))
+}
+
+func (l Label) String() string {
+	return l.rep().skv
+}
+
+func (l Label) Source() string { return l.rep().source() }
+func (l Label) Key() string    { return l.rep().key() }
+func (l Label) Value() string  { return l.rep().value() }
+
+func (l Label) Equal(other Label) bool {
+	return l == other
+}
+
+type jsonLabel struct {
+	Key    string `json:"key"`
+	Value  string `json:"value,omitempty"`
+	Source string `json:"source"`
+}
+
+func (l Label) MarshalJSON() ([]byte, error) {
+	return json.Marshal(jsonLabel{
+		l.Key(), l.Value(), l.Source(),
+	})
+}
+
+func (l *Label) UnmarshalJSON(b []byte) error {
+	var jl jsonLabel
+	err := json.Unmarshal(b, &jl)
+	if err != nil {
+		// If parsing of the full representation failed then try the short
+		// form in the format:
+		//
+		// [SOURCE:]KEY[=VALUE]
+		var aux string
+
+		if err := json.Unmarshal(b, &aux); err != nil {
+			return fmt.Errorf("decode of Label as string failed: %w", err)
+		}
+
+		if aux == "" {
+			return fmt.Errorf("invalid Label: Failed to parse %s as a string", b)
+		}
+
+		*l = ParseLabel(aux)
+		return nil
+	} else {
+		if jl.Key == "" {
+			return fmt.Errorf("invalid Label: '%s' does not contain label key", b)
+		}
+
+		*l = NewLabel(jl.Key, jl.Value, jl.Source)
+	}
+
+	if jl.Source == LabelSourceCIDR {
+		c, err := LabelToPrefix(l.Key())
+		if err == nil {
+			*l = NewCIDRLabel(jl.Key, jl.Value, jl.Source, &c)
+		} else {
+			return fmt.Errorf("failed to parse CIDR label: invalid prefix: %w", err)
+		}
+	} else {
+		*l = NewLabel(jl.Key, jl.Value, jl.Source)
+	}
+
+	return nil
+}
+
+func (l Label) rep() labelRep {
+	return (unique.Handle[labelRep])(l).Value()
+}
+
+// labelRep is the internal unique'd representation for a label.
+type labelRep struct {
+	// skv is the concatenated source, key and value. It is the direct
+	// string representation for the label: "Source:Key=Value".
+	// We're optimizing for low runtime memory usage and number of allocated
+	// objects by concatenating them.
+	skv string
+
+	// kpos is the position of the key in skv
+	kpos uint16
+
+	// vpos is the position of the value in skv
+	vpos uint16
+
+	// optimization for CIDR prefixes
+	cidr *netip.Prefix
+}
+
+func (rep labelRep) source() string {
```

Comment:
Is it allowed to have an empty source? Perhaps not in a serialized form, but certainly on the "ui" side, sometimes we need to represent labels without a source.